kernel/kernel-api/src/main/java/io/delta/kernel/internal/CommitActionsImpl.java

```diff
@@ -0,0 +1,196 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.internal;
+
+import static java.util.Objects.requireNonNull;
+
+import io.delta.kernel.CommitActions;
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.data.ColumnarBatch;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.internal.actions.Protocol;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.replay.ActionWrapper;
+import io.delta.kernel.internal.replay.ActionsIterator;
+import io.delta.kernel.internal.tablefeatures.TableFeatures;
+import io.delta.kernel.internal.util.FileNames;
+import io.delta.kernel.internal.util.Preconditions;
+import io.delta.kernel.types.StructField;
+import io.delta.kernel.types.StructType;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.FileStatus;
+import io.delta.kernel.utils.PeekableIterator;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Implementation of {@link CommitActions}.
+ *
+ * <p>This implementation owns the commit file and supports multiple calls to {@link #getActions()}.
+ * The first call reuses initially-read data to avoid double I/O, while subsequent calls re-read the
+ * commit file for memory efficiency.
+ *
+ * <p><b>Resource Management:</b>
+ *
+ * <ul>
+ *   <li>Calling {@link #getActions()} transfers resource ownership to the returned iterator.
+ *   <li>Callers MUST close the returned iterator (preferably using try-with-resources) to release
+ *       file handles and other resources.
+ *   <li>If {@link #getActions()} is never called, internal resources will be released when the
+ *       object is garbage collected (though explicit cleanup via {@link #getActions()} followed by
+ *       iterator closure is preferred).
+ * </ul>
+ */
+public class CommitActionsImpl implements CommitActions {
+
+  private final Engine engine;
+  private final FileStatus commitFile;
+  private final StructType readSchema;
+  private final String tablePath;
+  private final boolean shouldDropProtocolColumn;
+  private final boolean shouldDropCommitInfoColumn;
+  private final long version;
+  private final long timestamp;
+
+  /**
+   * Peekable iterator for ActionWrappers. Supports peeking at the first element for metadata
+   * extraction. Used for the first getActions() call, then set to null.
```

Comment:
I don't see it set to null I think this is not true? Can you update the expectations on the state of this variable?